src/run-state-uploader.ts

```diff
@@ -0,0 +1,57 @@
+import { RunState } from './types/run';
+import { UploadInfo } from './types/internal';
+
+export const RUN_STATE_UPLOAD_PERIOD = 10000;
+
+export interface Uploader {
+    getUploadInfo (uploadEntityId: string): Promise<UploadInfo>;
+    upload (uploadUrl: string, uploadEntity: Buffer): Promise<void>;
+}
+
+export interface RunStateProvider {
+    getRunState(): Partial<RunState>;
+}
+
+function skipEmptyValues (key, value) {
+    return value instanceof RegExp ? value.toString() : value;
+}
+
+export class RunStateUploader {
+    private _runStateProvider: RunStateProvider;
+    private _uploader: Uploader;
+    private _intervalId: NodeJS.Timeout;
+    private _uploadInfo: UploadInfo;
+    private _runStateUploadPeriod: number;
+
+    async _uploadRunState () {
+        const runState = this._runStateProvider.getRunState();
+
+        const buffer = Buffer.from(JSON.stringify(runState, skipEmptyValues));
+
+        this._uploader.upload(this._uploadInfo.uploadUrl, buffer);
+    }
+
+    constructor (uploader: Uploader, runStateProvider: RunStateProvider, runStateUploadPeriod = RUN_STATE_UPLOAD_PERIOD) {
+        this._uploader = uploader;
+        this._runStateProvider = runStateProvider;
+        this._runStateUploadPeriod = runStateUploadPeriod;
+
+        this._uploadRunState = this._uploadRunState.bind(this);
+    }
+
+    async start (runId: string) {
```

Comment:
If speaking about this class in isolation, I would say that we should add a way to prevent the bad consequences of calling `start()` twice.

Currently, it looks like it will begin a new cycle of uploading the same run state provider's data.

Reply:
Fixed